synapse/storage/data_stores/main/__init__.py

```diff
@@ -518,35 +521,39 @@ def get_users_paginate(
             guests (bool): whether to in include guest users
             deactivated (bool): whether to include deactivated users
         Returns:
-            defer.Deferred: resolves to list[dict[str, Any]]
+            defer.Deferred: resolves to list[dict[str, Any]], int
         """
-        name_filter = {}
+        filters = []
+        args = []
+
         if name:
-            name_filter["name"] = "%" + name + "%"
+            filters.append("name LIKE ?")
+            args.append("%" + name + "%")
 
-        attr_filter = {}
         if not guests:
-            attr_filter["is_guest"] = 0
+            filters.append("is_guest = 0")
+
         if not deactivated:
-            attr_filter["deactivated"] = 0
+            filters.append("deactivated = 0")
 
-        return self.db.simple_select_list_paginate(
-            desc="get_users_paginate",
-            table="users",
-            orderby="name",
-            start=start,
-            limit=limit,
-            filters=name_filter,
-            keyvalues=attr_filter,
-            retcols=[
-                "name",
-                "password_hash",
-                "is_guest",
-                "admin",
-                "user_type",
-                "deactivated",
-            ],
+        where_clause = "WHERE " + " AND ".join(filters) if len(filters) > 0 else ""
+
+        sql = "SELECT COUNT(*) as total_users FROM users %s" % (where_clause)
+        count = yield self.db.execute("count_users", self.db.cursor_to_dict, sql, *args)
```

Comment:
We tend to use `self.db.runInteraction` to run queries, which includes tracking statistics of how long the query took among other things which are useful for server admins to debug long-running transactions. For example usage, see here: https://github.com/matrix-org/synapse/blob/0ab5853ec93a75622e847aaf24efa91ae0037301/synapse/storage/data_stores/main/event_federation.py#L196-L220

Reply:
OK, I will try to use integrate this API.

Reply:
@anoadragon453 should we run both queries in the same runInteraction() call?

Reply:
Yes ideally, to make sure nothing changes between counting and getting the data out.

Reply:
While re-implementing this function, I found that `self.db.execute` actually calls `runInteraction()` internally. So why do I need to make this explicit?

Reply:
Ah, you're correct yes. It would still be good to run the queries in a single transaction for idempotency.

Reply:
Do you have an example how to do this?

Reply:
@anoadragon453 ping

Reply:
@awesome-manuel Sorry for the insanely long delay on this, got lost in my github notifications.

Here's a good example:

https://github.com/matrix-org/synapse/blob/30e9adf32f7db9aa44693876d33692bf4590d363/synapse/storage/data_stores/main/account_data.py#L193-L213

You create a `*_txn`  function inline, then pass it to `self.db.runInteraction()`. I'm not sure whether `self.db.execute` would suffice in this case though, however you can run `cursor_to_dict` inside the `txn` function before returning.

Reply:
@anoadragon453 Thanks for the hint! That has helped a lot.